File: src/utils/services_pdf_builder.py
#!/usr/bin/env python3
"""
Build PDF presentation for services catalog export.
"""

import io
from pathlib import Path

from reportlab.lib import colors
from reportlab.lib.enums import TA_CENTER, TA_LEFT, TA_RIGHT
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet
from reportlab.lib.units import mm
from reportlab.platypus import Paragraph, SimpleDocTemplate, Spacer, Table, TableStyle

from utils.pdf_report_builder import register_report_font
from utils.service_units import is_percent_unit
from utils.services_excel_builder import _group_services_by_paragraph, resolve_excel_path


def resolve_pdf_path(file_url):
    """Resolve local PDF path from plain path or file:// URL."""
    path = resolve_excel_path(file_url)
    if not path:
        return ""

    if path.lower().endswith(".xlsx"):
        path = path[:-5]

    if not path.lower().endswith(".pdf"):
        path += ".pdf"
    return path


def _format_price_for_pdf(cents, unit):
    value = int(cents or 0)
    if is_percent_unit(unit):
        if value % 100 == 0:
            return str(value // 100)
        text = f"{value / 100:.2f}".rstrip("0").rstrip(".")
        return text

    return f"{value / 100:.2f} \u20bd"


class ServicesPdfBuilder:
    """Build grouped services catalog PDF presentation."""

    _TABLE_COL_WIDTHS = [105 * mm, 35 * mm, 34 * mm]

    def __init__(self, font_path=""):
        self._font_name = register_report_font(font_path)
        self._styles = self._build_styles()

    def _build_styles(self):
        styles = getSampleStyleSheet()
        base_kwargs = {"fontName": self._font_name}

        return {
            "title": ParagraphStyle(
                "ServicesPdfTitle",
                parent=styles["Heading1"],
                alignment=TA_CENTER,
                fontSize=14,
                leading=18,
                spaceAfter=8,
                **base_kwargs,
            ),
            "section": ParagraphStyle(
                "ServicesPdfSection",
                parent=styles["Heading2"],
                fontSize=11,
                leading=15,
                spaceBefore=6,
                spaceAfter=4,
                **base_kwargs,
            ),
            "table_cell": ParagraphStyle(
                "ServicesPdfTableCell",
                parent=styles["Normal"],
                alignment=TA_LEFT,
                fontSize=10,
                leading=13,
                **base_kwargs,
            ),
            "table_price": ParagraphStyle(
                "ServicesPdfTablePrice",
                parent=styles["Normal"],
                alignment=TA_RIGHT,
                fontSize=10,
                leading=13,
                **base_kwargs,
            ),
            "table_unit": ParagraphStyle(
                "ServicesPdfTableUnit",
                parent=styles["Normal"],
                alignment=TA_CENTER,
                fontSize=10,
                leading=13,
                **base_kwargs,
            ),
        }

    def _escape(self, value):
        text = str(value or "")
        return (
            text.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
        )

    def _service_description(self, name, note):
        name_text = self._escape(str(name or "").strip())
        note_text = self._escape(str(note or "").strip())
        if note_text:
            return f"{name_text}<br/>{note_text}"
        return name_text

    def _service_table_row(self, service):
        name = str(service.get("name") or "").strip()
        note = str(service.get("note") or "").strip()
        unit = str(service.get("unit") or "").strip()
        price_text = _format_price_for_pdf(service.get("price", 0), unit)

        return [
            Paragraph(self._service_description(name, note), self._styles["table_cell"]),
            Paragraph(self._escape(price_text), self._styles["table_price"]),
            Paragraph(self._escape(unit), self._styles["table_unit"]),
        ]

    def _table_style(self):
        return TableStyle([
            ("GRID", (0, 0), (-1, -1), 0.5, colors.grey),
            ("VALIGN", (0, 0), (-1, -1), "TOP"),
            ("TOPPADDING", (0, 0), (-1, -1), 4),
            ("BOTTOMPADDING", (0, 0), (-1, -1), 4),
            ("LEFTPADDING", (0, 0), (-1, -1), 4),
            ("RIGHTPADDING", (0, 0), (-1, -1), 4),
        ])

    def build_pdf_bytes(self, services, title="Презентация услуг"):
        buffer = io.BytesIO()
        doc = SimpleDocTemplate(
            buffer,
            pagesize=A4,
            leftMargin=18 * mm,
            rightMargin=18 * mm,
            topMargin=16 * mm,
            bottomMargin=16 * mm,
            title=title,
        )

        story = [Paragraph(self._escape(title), self._styles["title"])]

        for paragraph, items in _group_services_by_paragraph(services):
            if paragraph:
                story.append(Paragraph(self._escape(paragraph), self._styles["section"]))

            if not items:
                continue

            data = [self._service_table_row(service) for service in items]
            table = Table(data, colWidths=self._TABLE_COL_WIDTHS)
            table.setStyle(self._table_style())
            story.append(table)
            story.append(Spacer(1, 4 * mm))

        doc.build(story)
        return buffer.getvalue()

    def save_pdf(self, file_path, services, title="Презентация услуг"):
        path = Path(file_path)
        path.parent.mkdir(parents=True, exist_ok=True)
        pdf_bytes = self.build_pdf_bytes(services, title=title)
        path.write_bytes(pdf_bytes)
        return str(path.resolve())


def export_services_pdf(file_url, services=None, title="Презентация услуг"):
    """Write services catalog presentation PDF at the given path or URL."""
    path = resolve_pdf_path(file_url)
    if not path:
        raise ValueError("Services presentation path is empty")

    if services is None:
        from utils.services_db import load_services
        services = load_services()

    return ServicesPdfBuilder().save_pdf(path, services, title=title)

File: cpp/utils/ServicesPdfBuilder.cpp
#include "ServicesPdfBuilder.h"

#include "PdfReportBuilder.h"
#include "ServiceUnits.h"
#include "ServicesDatabase.h"
#include "ServicesExcelBuilder.h"

#include <QBuffer>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QPdfWriter>
#include <QTextDocument>
#include <QUrl>
#include <algorithm>

#include <stdexcept>

namespace workorder
{

namespace
{

using ParagraphGroup = QPair<QString, StringMapList>;

QString formatPriceForPdf(int cents, const QString &unit)
{
    const int value = cents;
    if(ServiceUnits::isPercentUnit(unit))
    {
        if(value % 100 == 0)
            return QString::number(value / 100);
        QString text = QString::number(value / 100.0, 'f', 2);
        while(text.endsWith('0'))
            text.chop(1);
        if(text.endsWith('.'))
            text.chop(1);
        return text;
    }

    return QStringLiteral("%1 \u20bd").arg(QString::number(value / 100.0, 'f', 2));
}

QString escapeHtml(const QString &value)
{
    QString text = value;
    text.replace('&', "&amp;");
    text.replace('<', "&lt;");
    text.replace('>', "&gt;");
    return text;
}

QVector<ParagraphGroup> groupServicesByParagraph(const StringMapList &services)
{
    QMap<QString, StringMapList> grouped;
    for(const QVariant &item : services)
    {
        const StringMap service = item.toMap();
        grouped[service.value("paragraph").toString().trimmed()].append(service);
    }

    QStringList paragraphs = grouped.keys();
    std::sort(paragraphs.begin(), paragraphs.end(), [](const QString &a, const QString &b)
    {
        if(a.isEmpty())
            return false;
        if(b.isEmpty())
            return true;
        return a.compare(b, Qt::CaseInsensitive) < 0;
    });

    QVector<ParagraphGroup> result;
    for(const QString &paragraph : paragraphs)
    {
        StringMapList items = grouped.value(paragraph);
        std::sort(items.begin(), items.end(), [](const QVariant &left, const QVariant &right)
        {
            return left.toMap().value("name").toString().compare(
                right.toMap().value("name").toString(), Qt::CaseInsensitive) < 0;
        });
        result.append({paragraph, items});
    }
    return result;
}

QString buildServicesHtml(const StringMapList &services, const QString &title, const QString &fontPath)
{
    const QString fontFamily = PdfReportBuilder::registerReportFont(fontPath);
    const QString resolvedFontPath = PdfReportBuilder::resolveReportFontPath(fontPath);

    QString html = QStringLiteral("<html><head><meta charset=\"utf-8\"/>");
    if(!resolvedFontPath.isEmpty())
    {
        html += QStringLiteral("<style>@font-face { font-family: '%1'; src: url('%2'); }"
                               "body, table { font-family: '%1', sans-serif; font-size: 10pt; }</style>")
            .arg(escapeHtml(fontFamily), QUrl::fromLocalFile(resolvedFontPath).toString());
    }
    html += QStringLiteral("</head><body>");
    html += QStringLiteral("<h2 style=\"text-align:center;\">%1</h2>").arg(escapeHtml(title));

    for(const ParagraphGroup &group : groupServicesByParagraph(services))
    {
        if(!group.first.isEmpty())
            html += QStringLiteral("<h3>%1</h3>").arg(escapeHtml(group.first));

        if(group.second.isEmpty())
            continue;

        html += QStringLiteral("<table border=\"1\" cellspacing=\"0\" cellpadding=\"4\" width=\"100%\">");
        for(const QVariant &item : group.second)
        {
            const StringMap service = item.toMap();
            const QString name = service.value("name").toString().trimmed();
            const QString note = service.value("note").toString().trimmed();
            const QString unit = service.value("unit").toString().trimmed();
            const QString priceText = formatPriceForPdf(service.value("price").toInt(), unit);

            QString description = escapeHtml(name);
            if(!note.isEmpty())
                description += QStringLiteral("<br/>%1").arg(escapeHtml(note));

            html += QStringLiteral("<tr><td>%1</td><td align=\"right\">%2</td><td align=\"center\">%3</td></tr>")
                .arg(description, escapeHtml(priceText), escapeHtml(unit));
        }
        html += QStringLiteral("</table><br/>");
    }

    html += QStringLiteral("</body></html>");
    return html;
}

} // namespace

ServicesPdfBuilder::ServicesPdfBuilder(const QString &fontPath)
    : m_fontPath(fontPath)
{
}

QString ServicesPdfBuilder::resolvePdfPath(const QString &fileUrl)
{
    QString path = ServicesExcelBuilder::resolveExcelPath(fileUrl);
    if(path.isEmpty())
        return QString();

    if(path.endsWith(".xlsx", Qt::CaseInsensitive))
        path.chop(5);

    if(!path.endsWith(".pdf", Qt::CaseInsensitive))
        path += ".pdf";

    return path;
}

QByteArray ServicesPdfBuilder::buildPdfBytes(const StringMapList &services, const QString &title) const
{
    const QString resolvedTitle = title.isEmpty() ? QStringLiteral("Презентация услуг") : title;

    QTextDocument document;
    document.setHtml(buildServicesHtml(services, resolvedTitle, m_fontPath));

    QByteArray bytes;
    QBuffer buffer(&bytes);
    buffer.open(QIODevice::WriteOnly);

    QPdfWriter writer(&buffer);
    writer.setPageSize(QPageSize(QPageSize::A4));
    writer.setPageMargins(QMarginsF(16, 16, 16, 16), QPageLayout::Millimeter);
    document.print(&writer);

    return bytes;
}

QString ServicesPdfBuilder::savePdf(const QString &filePath, const StringMapList &services,
                                      const QString &title) const
{
    QDir().mkpath(QFileInfo(filePath).absolutePath());
    const QByteArray pdfBytes = buildPdfBytes(services, title);

    QFile file(filePath);
    if(!file.open(QIODevice::WriteOnly))
        return QString();

    file.write(pdfBytes);
    return QFileInfo(filePath).absoluteFilePath();
}

QString exportServicesPdf(const QString &fileUrl, const StringMapList &services, const QString &title)
{
    const QString path = ServicesPdfBuilder::resolvePdfPath(fileUrl);
    if(path.isEmpty())
        throw std::runtime_error("Services presentation path is empty");

    StringMapList data = services;
    if(data.isEmpty())
        data = ServicesDatabase::loadServices();

    return ServicesPdfBuilder().savePdf(path, data, title);
}

} // namespace workorder
